// common/responseTool.js
var responseTool = {
    /*返回服务端异常*/
    sendServerException: function (res, mess) {
        res.send(500, mess || '服务端错误请查看后台日志');
    },
    /*返回拒绝访问*/
    sendDecline: function (res, mess) {
        res.send(405, mess || '服务端拒绝访问');
    },
    /*返回数据更新、上传、下载成功*/
    sendSuccess: function (res, param) {
        param = param || {};
        res.set('Content-Type', 'text/json;charset=utf-8');
        res.send(param);
    },
    /*重定向*/
    sendRedirect: function (res, url) {
        res.send(302, url);
    },
    /*登录请求的响应
    *result:{
        err:'',     //异常信息
        isLogin:true,  //为true时代表登录成功，否则登录失败
        loginFailedType:1,  //登录失败原因，2 用户名不存在  3 密码错误 1 用户名或密码错误
        puser:{}            //用户信息
    *}
    */
    loginResponse: function (req, res, result) {
        var err = result.err;
        if (err) {
            console.error('登录err：' + (err.stack || JSON.stringify(err)));
            return this.renderLogin(res, 1);
        }
        var isLogin = result.isLogin || false;
        if (isLogin) {
            req.session.puser = result.puser || {};
            res.redirect('/');
            return;
        }
        this.renderLogin(res, result.loginFailedType);
    },
    /*渲染主页面*/
    renderMain: function (res, user) {
        res.render('layout', { host: commonLibUrl, user: user || {} });
    },
    /*渲染登录页面*/
    renderLogin: function (res, err) {
        res.render('login', { host: commonLibUrl, err: err });
    },
    /*跳转到首页*/
    redirectMain: function (res) {
        res.redirect('/');
    }
};

module.exports = responseTool;
